File: frontend/src/api/query.ts
/** Query API client — consumes /api/v1/query endpoints from Plan 02. */

// --- Response types ---

export interface WorkflowQueryResponse {
  id: string;
  template_name: string;
  template_version: number;
  state: string;
  started_by: string | null;
  started_at: string | null;
  completed_at: string | null;
  active_activity: string | null;
}

export interface WorkItemQueryResponse {
  id: string;
  activity_name: string;
  workflow_name: string;
  workflow_id: string;
  assignee: string | null;
  state: string;
  priority: number;
  created_at: string;
}

export interface DocumentQueryResponse {
  id: string;
  title: string;
  lifecycle_state: string | null;
  current_version: string;
  author: string | null;
  created_by: string | null;
  updated_at: string;
  content_type: string;
}

export interface PaginationMeta {
  total: number;
  skip: number;
  limit: number;
  total_pages: number;
}

export interface PaginatedResponse<T> {
  data: T[];
  meta: PaginationMeta;
}

// --- Auth helpers (mirrors templates.ts) ---

function authHeaders(): HeadersInit {
  const token = localStorage.getItem("token");
  return token ? { Authorization: `Bearer ${token}` } : {};
}

async function apiFetch<T>(url: string): Promise<T> {
  const res = await fetch(url, {
    headers: {
      "Content-Type": "application/json",
      ...authHeaders(),
    },
  });
  if (!res.ok) {
    const body = await res.text();
    throw new Error(`API error ${res.status}: ${body}`);
  }
  return res.json() as Promise<T>;
}

// --- Param builder ---

function buildUrl(
  base: string,
  params: Record<string, string | number | undefined | null>,
): string {
  const url = new URL(base, window.location.origin);
  for (const [key, value] of Object.entries(params)) {
    if (value !== undefined && value !== null && value !== "") {
      url.searchParams.set(key, String(value));
    }
  }
  return url.pathname + url.search;
}

// --- Query functions ---

export interface WorkflowQueryParams {
  template_id?: string;
  state?: string;
  date_from?: string;
  date_to?: string;
  started_by?: string;
  skip?: number;
  limit?: number;
}

export async function queryWorkflows(
  params: WorkflowQueryParams,
): Promise<PaginatedResponse<WorkflowQueryResponse>> {
  const url = buildUrl("/api/v1/query/workflows", params);
  return apiFetch<PaginatedResponse<WorkflowQueryResponse>>(url);
}

export interface WorkItemQueryParams {
  assignee_id?: string;
  state?: string;
  workflow_id?: string;
  priority?: number;
  skip?: number;
  limit?: number;
}

export async function queryWorkItems(
  params: WorkItemQueryParams,
): Promise<PaginatedResponse<WorkItemQueryResponse>> {
  const url = buildUrl("/api/v1/query/work-items", {
    ...params,
    priority: params.priority,
  });
  return apiFetch<PaginatedResponse<WorkItemQueryResponse>>(url);
}

export interface DocumentQueryParams {
  lifecycle_state?: string;
  metadata_key?: string;
  metadata_value?: string;
  version?: string;
  skip?: number;
  limit?: number;
}

export async function queryDocuments(
  params: DocumentQueryParams,
): Promise<PaginatedResponse<DocumentQueryResponse>> {
  const url = buildUrl("/api/v1/query/documents", params);
  return apiFetch<PaginatedResponse<DocumentQueryResponse>>(url);
}

// --- Utility: fetch templates and users for filter dropdowns ---

export interface TemplateSummary {
  id: string;
  name: string;
}

export interface UserSummary {
  id: string;
  username: string;
}

export async function fetchTemplatesForFilter(): Promise<TemplateSummary[]> {
  const res = await apiFetch<{ data: TemplateSummary[] }>("/api/v1/templates/");
  return res.data;
}

export async function fetchUsersForFilter(): Promise<UserSummary[]> {
  const res = await apiFetch<{ data: UserSummary[] }>("/api/v1/users/");
  return res.data;
}

// --- Badge color helper ---

export function getStateBadgeClass(state: string): string {
  const map: Record<string, string> = {
    running: "bg-green-100 text-green-700",
    halted: "bg-amber-100 text-amber-700",
    finished: "bg-blue-100 text-blue-700",
    failed: "bg-red-100 text-red-700",
    dormant: "bg-secondary text-secondary-foreground",
    available: "bg-green-100 text-green-700",
    acquired: "bg-blue-100 text-blue-700",
    complete: "bg-gray-100 text-gray-700",
    completed: "bg-gray-100 text-gray-700",
    rejected: "bg-red-100 text-red-700",
    suspended: "bg-amber-100 text-amber-700",
    draft: "bg-gray-100 text-gray-700",
    review: "bg-amber-100 text-amber-700",
    approved: "bg-green-100 text-green-700",
    archived: "bg-blue-100 text-blue-700",
  };
  return map[state.toLowerCase()] || "bg-secondary text-secondary-foreground";
}

File: src/app/routers/query.py
"""Admin query endpoints for workflows, work items, and documents."""
import math
import uuid
from datetime import datetime

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.dependencies import get_current_active_admin
from app.models.user import User
from app.schemas.common import EnvelopeResponse, PaginationMeta
from app.schemas.query import (
    DocumentQueryResponse,
    WorkflowQueryResponse,
    WorkItemQueryResponse,
)
from app.services import query_service

router = APIRouter(prefix="/query", tags=["query"])


@router.get("/workflows", response_model=EnvelopeResponse[list[WorkflowQueryResponse]])
async def query_workflows_endpoint(
    template_id: uuid.UUID | None = Query(None),
    state: str | None = Query(None),
    date_from: datetime | None = Query(None),
    date_to: datetime | None = Query(None),
    started_by: uuid.UUID | None = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_admin),
):
    """Query workflow instances with multi-criteria filtering. Admin only."""
    items, total_count = await query_service.query_workflows(
        db,
        template_id=template_id,
        state=state,
        date_from=date_from,
        date_to=date_to,
        started_by=started_by,
        skip=skip,
        limit=limit,
    )
    return EnvelopeResponse(
        data=items,
        meta=PaginationMeta(
            page=(skip // limit) + 1,
            page_size=limit,
            total_count=total_count,
            total_pages=math.ceil(total_count / limit) if limit > 0 else 0,
        ),
    )


@router.get("/work-items", response_model=EnvelopeResponse[list[WorkItemQueryResponse]])
async def query_work_items_endpoint(
    assignee_id: uuid.UUID | None = Query(None),
    state: str | None = Query(None),
    workflow_id: uuid.UUID | None = Query(None),
    priority: int | None = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_admin),
):
    """Query work items with multi-criteria filtering. Admin only."""
    items, total_count = await query_service.query_work_items(
        db,
        assignee_id=assignee_id,
        state=state,
        workflow_id=workflow_id,
        priority=priority,
        skip=skip,
        limit=limit,
    )
    return EnvelopeResponse(
        data=items,
        meta=PaginationMeta(
            page=(skip // limit) + 1,
            page_size=limit,
            total_count=total_count,
            total_pages=math.ceil(total_count / limit) if limit > 0 else 0,
        ),
    )


@router.get("/documents", response_model=EnvelopeResponse[list[DocumentQueryResponse]])
async def query_documents_endpoint(
    lifecycle_state: str | None = Query(None),
    metadata_key: str | None = Query(None),
    metadata_value: str | None = Query(None),
    version: str | None = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_admin),
):
    """Query documents with multi-criteria filtering. Admin only."""
    items, total_count = await query_service.query_documents(
        db,
        lifecycle_state=lifecycle_state,
        metadata_key=metadata_key,
        metadata_value=metadata_value,
        version=version,
        skip=skip,
        limit=limit,
    )
    return EnvelopeResponse(
        data=items,
        meta=PaginationMeta(
            page=(skip // limit) + 1,
            page_size=limit,
            total_count=total_count,
            total_pages=math.ceil(total_count / limit) if limit > 0 else 0,
        ),
    )
